File: Core/SyncOperation.vb
Imports System.IO

Public Enum FileOperation
    None
    Add
    Replace
    Remove
End Enum

Public Class SyncOperation

    Private _sourceFilePath As String
    Private _targetFilePath As String
    Private _relativeFilePath As String
    Private _operation As FileOperation
    Private _exemption As SyncTaskExemption

    Public Event SyncOperationStarted(ByVal operation As SyncOperation)
    Public Event SyncOperationProgressed(ByVal bytesTotal As Long, ByVal bytesTransferred As Long)
    Public Event SyncOperationFinished(ByVal operation As SyncOperation, ByVal e As SyncOperationFinishedEventArgs)

    Friend Sub New()

    End Sub

#Region "PROPERTIES"

    Public Property SourceFilePath As String
        Get
            Return _sourceFilePath
        End Get
        Friend Set(value As String)
            _sourceFilePath = value
        End Set
    End Property

    Public Property TargetFilePath As String
        Get
            Return _targetFilePath
        End Get
        Friend Set(value As String)
            _targetFilePath = value
        End Set
    End Property

    Public Property RelativeFilePath As String
        Get
            Return _relativeFilePath
        End Get
        Friend Set(value As String)
            _relativeFilePath = value
        End Set
    End Property

    Public Property Operation As FileOperation
        Get
            Return _operation
        End Get
        Friend Set(value As FileOperation)
            _operation = value
        End Set
    End Property

    Public Property Exemption As SyncTaskExemption
        Get
            Return _exemption
        End Get
        Friend Set(value As SyncTaskExemption)
            _exemption = value
        End Set
    End Property

#End Region

#Region "METHODS"

    Private Function CopyProgress(ByVal totalFileSize As Long, _
                                  ByVal totalBytesTransferred As Long, _
                                  ByVal streamSize As Long, _
                                  ByVal streamBytesTransferred As Long, _
                                  ByVal streamNumber As UInteger,
                                  ByVal callbackReason As Win32API.CopyProgressCallbackReason, _
                                  ByVal sourceFile As IntPtr,
                                  ByVal destinationFile As IntPtr,
                                  ByVal data As IntPtr) As Win32API.CopyProgressResult
        RaiseEvent SyncOperationProgressed(totalFileSize, totalBytesTransferred)
        Return Win32API.CopyProgressResult.PROGRESS_CONTINUE
    End Function

    Private Sub CreateParentDirectory()
        Dim sourceParentDirectory As String = Path.GetDirectoryName(Me.SourceFilePath)
        Dim targetParentDirectory As String = Path.GetDirectoryName(Me.TargetFilePath)
        If Not Directory.Exists(targetParentDirectory) Then
            Directory.CreateDirectory(targetParentDirectory)
            File.SetAttributes(targetParentDirectory, File.GetAttributes(sourceParentDirectory))
        End If
    End Sub

    Private Sub DeleteParentDirectory()
        Dim parentDirectory As New DirectoryInfo(Path.GetDirectoryName(Me.TargetFilePath))
        If (Not parentDirectory.EnumerateDirectories().Any()) AndAlso (Not parentDirectory.EnumerateFiles().Any()) Then
            parentDirectory.Delete(False)
        End If
    End Sub

    Public Sub Execute(ByRef stopRequested As Boolean)
        Try
            RaiseEvent SyncOperationStarted(Me)
            Select Case Me.Operation
                Case FileOperation.Add
                    CreateParentDirectory()
                    Win32API.CopyFileEx(Me.SourceFilePath, Me.TargetFilePath, AddressOf CopyProgress, Nothing, stopRequested, Nothing)
                Case FileOperation.Replace
                    Win32API.CopyFileEx(Me.SourceFilePath, Me.TargetFilePath, AddressOf CopyProgress, Nothing, stopRequested, Nothing)
                Case FileOperation.Remove
                    File.SetAttributes(Me.TargetFilePath, FileAttributes.Normal)
                    File.Delete(Me.TargetFilePath)
                    If Path.GetDirectoryName(Me.TargetFilePath) <> Me.TargetFilePath.Replace(Path.DirectorySeparatorChar & Me.RelativeFilePath, Space(0)) Then
                        DeleteParentDirectory()
                    End If
            End Select
            If Not stopRequested Then
                RaiseEvent SyncOperationFinished(Me, New SyncOperationFinishedEventArgs())
            Else
                RaiseEvent SyncOperationFinished(Me, New SyncOperationFinishedEventArgs(True))
            End If
        Catch ex As Exception
            RaiseEvent SyncOperationFinished(Me, New SyncOperationFinishedEventArgs(False, True, ex))
        End Try
    End Sub

#End Region

End Class

Public Class SyncOperationFinishedEventArgs

    Friend Sub New(Optional ByVal canceled As Boolean = False, _
                   Optional ByVal failed As Boolean = False, _
                   Optional ByVal failureReason As Exception = Nothing)
        Me.Canceled = canceled
        Me.Failed = failed
        Me.FailureReason = failureReason
    End Sub

    Public Property Canceled As Boolean
    Public Property Failed As Boolean
    Public Property FailureReason As Exception

End Class

File: BP.Unify.Core/SyncOperation.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace BP.Unify.Core
{
	public enum FileOperation
	{
		None = 0,
		Add = 1,
		Replace = 2,
		Remove = 3
	}

	public class SyncOperation
	{
		public delegate void SyncOperationStartedHandler(SyncOperation operation);
		public delegate void SyncOperationProgressedHandler(long bytesTotal, long bytesTransferred);
		public delegate void SyncOperationFinishedHandler(SyncOperation operation, SyncOperationFinishedEventArgs e);

		public event SyncOperationStartedHandler SyncOperationStarted;
		public event SyncOperationProgressedHandler SyncOperationProgressed;
		public event SyncOperationFinishedHandler SyncOperationFinished;

		internal SyncOperation()
		{

		}

		#region PROPERTIES

		public string SourceFilePath { get; set; }
		public string TargetFilePath { get; set; }
		public string RelativeFilePath { get; set; }
		public FileOperation Operation { get; set; }
		public SyncTaskExemption Exemption { get; set; }

		#endregion

		#region PUBLIC METHODS

		public void Execute(ref bool stopRequested)
		{
			try
			{
				this.SyncOperationStarted(this);
				switch(this.Operation)
				{
					case FileOperation.Add:
						CreateParentDirectory();
						Win32API.CopyFileEx(this.SourceFilePath, this.TargetFilePath, CopyProgress, null, ref stopRequested, null);
					case FileOperation.Replace:
						Win32API.CopyFileEx(this.SourceFilePath, this.TargetFilePath, CopyProgress, null, ref stopRequested, null);
					case FileOperation.Remove:
						File.SetAttributes(this.TargetFilePath, FileAttributes.Normal);
						File.Delete(this.TargetFilePath);
						if (Path.GetDirectoryName(this.TargetFilePath) != this.TargetFilePath.Replace(Path.DirectorySeparatorChar + this.RelativeFilePath, ""))
						{
							DeleteParentDirectory();
						}
				}
				if (!stopRequested)
				{ this.SyncOperationFinished(this, new SyncOperationFinishedEventArgs()); }
				else
				{ this.SyncOperationFinished(this, new SyncOperationFinishedEventArgs(true)); }		
			}
			catch (Exception ex)
			{
				this.SyncOperationFinished(this, new SyncOperationFinishedEventArgs(false, true, ex));
			}
		}

		#endregion

		#region PRIVATE METHODS

		private Win32API.CopyProgressResult CopyProgress(long totalFileSize, long totalBytesTransferred, long streamSize, long streamBytesTransferred, uint streamNumber, Win32API.CopyProgressCallbackReason callbackReason, IntPtr sourceFile, IntPtr destinationFile, IntPtr data)
		{
			this.SyncOperationProgressed(totalFileSize, totalBytesTransferred);
			return Win32API.CopyProgressResult.PROGRESS_CONTINUE;
		}

		private void CreateParentDirectory()
		{
			string sourceParentDirectory = Path.GetDirectoryName(this.SourceFilePath);
			string targetParentDirectory = Path.GetDirectoryName(this.TargetFilePath);
			if (!Directory.Exists(targetParentDirectory))
			{
				Directory.CreateDirectory(targetParentDirectory);
				File.SetAttributes(targetParentDirectory, File.GetAttributes(sourceParentDirectory));
			}
		}

		private void DeleteParentDirectory()
		{
			DirectoryInfo parentDirectory = new DirectoryInfo(Path.GetDirectoryName(this.TargetFilePath));
			if (!parentDirectory.EnumerateDirectories().Any() && !parentDirectory.EnumerateFiles().Any())
			{
				parentDirectory.Delete(false);
			}
		}

		#endregion
	}

	public class SyncOperationFinishedEventArgs
	{
		internal SyncOperationFinishedEventArgs(bool canceled = false, bool failed = false, Exception failureReason = null)
		{
			this.Canceled = canceled;
			this.Failed = failed;
			this.FailureReason = failureReason;
		}

		public bool Canceled { get; set; }
		public bool Failed { get; set; }
		public Exception FailureReason { get; set; }
	}
}
